### DISCUSSION TITLE
Automatically add items from repositories (projects beta)

### DISCUSSION THREAD

**Discussion by lfrancke** [community/community#5378] [Projects and Issues] [answered]
I'd like to have a project where all issues and PRs are automatically added so they can be triaged and tagged/labeled/moved accordingly.

I couldn't find a way to do this now and I also didn't find an existing discussion about this. It would be super helpful when there are lots of repositories to check.

**Comment by petewalker** [comment]
If you have a public repo (or action minutes to burn), you could create a workflow to do this for you. Would probably look something like:

https://docs.github.com/en/issues/trying-out-the-new-projects-experience/automating-projects#example-workflow

But I agree it would be good for this to be core functionality.

**Reply by lfrancke** [reply]
Thank you. That's a good point, I managed to miss those examples.

**Reply by troyschneringer** [reply]
I've tried following that documentation and consistently get a `Could not resolve to a node with the global id xxx` error when new `Issues` are created.  Has anyone had success creating a workflow that adds issues to a project?

**Reply by jenschude** [reply]
Yes. But you need a PAT with scope `repo`/`public_repo` and `write:org`

**Comment by Totktonada** [comment]
I filed another feature request, which may be considered as relevant in context of some use cases: https://github.com/github/feedback/discussions/5528.

Sure, both may be solved using a hand crafted automation, but it may require significant work if we're about an organization with a lot of repositories (if we're about per-repository action).

We can collect all issues event across the organization periodically and put them into a project using the API, but we should store a state somewhere and I worry about stability of such solution. It would be undesirable to miss several issues just because of a network or hardware failure.

**Comment by speg** [comment]
Just got beta access and this was the first thing I tried to do 😢 

To make matters worse, the search criteria doesn't seem to stick. So I would add `is:open` to filter only the open issues, add one from the suggestions list, and then the search would revert and I'd start seeing closed issues & PRs again.

**Comment by aspearman** [comment]
Automating how issues made their way into the project was one of the first things I looked into. 

I wonder if the "workflows" feature within projects -- which includes an "item added to project" default workflow -- can be extended to have an "item added to linked repository" as well, to trigger a workflow when an issue is added to a relevant repo. 

Then we could automate if/how an issue in a linked repo makes its way into the project.

**Comment by hu1buerger** [comment]
This could be problematic if the repo has multiple projects...
But pattern matching could work.

Kinda like "add all nee issues to this project"
and in another one "add all new issues where title matches "[core]""

Depending on the implementation both or the more specific one catches

**Comment by alexcnichols** [comment, marked as answer]
👋🏼 Improvements for auto adding issues to projects (beta), and more, are coming soon! Check out this roadmap item for more and subscribe to be notified if/when things change. https://github.com/github/roadmap/issues/286 :octocat: 

Let's keep the discussion going here but mentioning other (duplicate) discussions in case we need more about the use cases here:

- https://github.com/github/feedback/discussions/5713
- https://github.com/github/feedback/discussions/6833

**Reply by ynouri** [reply]
@alexcnichols I find the entire premise of requiring issues to be "added" to a project counter productive. It's not clear to me what use cases this feature is enabling (as opposed to have just all issues available by default). This is adding a lot of work and complexity (even through automation or workflows), especially in our use case where we just want to see all issues.

Filtering issues in a given board could easily be done by using a search / filter with a label for example. The process of "adding to a project" just becomes labeling issues - which can already be done in bulk. To make things easier, maybe a global filter at the project level (instead of the view level) would help.

**Reply by ynouri** [reply]
If it does enable certain classes of use cases, then maybe make it an opt-in feature, not the default. It seems much more intuitive to me to have all issues of your repo / org (see https://github.com/github/feedback/discussions/5528) in sync with your project by default (ie no need to introduce a workflow just to make sure you are not missing any issue)

**Reply by markcarroll** [reply]
here, here.  Why would I want to write a workflow to do this when it should just be there.   Let me choose which repos should be included and have all open issues from those automatically visible in the project.  Filters, sure! Give me filters, but having to automatically add them based on a workflow that never seemed to work right in the last incarnation, is just plain painful.

I love how this is coming along, it is already looking great, but *please* fix the issue adding process.

**Reply by luxaritas** [reply]
I'm assuming the intent is for usage as an epic, an alternative to a milestone, etc.  Not sure how that plays into the plans in github/roadmap#284. In most cases, I think it would be sufficient to define a project via filters (eg, all issues within some repo, all issues within a given milestone, all issues with a certain label, all issues in some epic, etc) - that could actually solve some confusion with what role a project has vs a milestone or other features. I do wonder if there are any use cases where adding issues to a project manually would make sense but not creating something like a milestone, label, or custom field and having the project track that.

**Reply by markcarroll** [reply]
I am hoping to use ours as a high level project prioritization tool.  Having a high level view over milestones/epics/features is the one thing that we miss having moved from Jira a while back.

**Reply by Alazoral** [reply]
@ynouri I agree entirely. Surely the default use case is 'view my repo's issues as a board'. It's so tedious to add each of them manually to a project, and seems designed to let issues fall into the cracks.

**Reply by tannerbaum** [reply]
Agreed with @ynouri and the others. In general I find it interesting that there doesn't seem to be a way to couple the projects beta to a single repository. I think in at least 90% of the cases we would use Projects at my company they would be exclusive to a single repository. So having to specify which repository to search in to add issues (which would be alleviated by this feature to be fair) is a bit cumbersome.

**Reply by RolfBippus** [reply]
The main feature of projects for us is to manage issues across multiple repos! For a single repo I would personally not necessarily use it, because I am not too much a fan of an additional project management tool for the sake of it and the repo already offers the issues and pull requests (although projects(beta) gives a better overview, agreed). But all the more, a simple setting to link repos and projects automatically is mandatory for us before continuing to extensively use projects(beta).

A poor man's solution could even be to at least be able to add a default project link to the issues and pull-request templates of a repo.

**Reply by michael-hawker** [reply]
Related to #6241 and #9764 follow-ups to this to have some more automated ways to pull in new features as well.

**Reply by qqpann** [reply]
Adding issues to project should be a default. **Please remove the five limitation**. 
And as others say, a project ready belongs to multiple projects. How about having an option to make a repository only belongs to one project. In such case it should not require any limitation.

**Comment by markcarroll** [comment]
Adding these here since my [other thread](https://github.com/github/feedback/discussions/6833) was marked as duplicate:

- Add all issues in a milestone to a project
- Add all child issues from a parent issue to a project
- Add option to bulk add issues to a project from the issues list screen (like you can assign, add labels etc) 
- Set the workflow such that all issues added to a repository can be automatically added to a project with filters so that all new issues of a certain criteria are added

**Comment by thomaseizinger** [comment]
For anyone looking for an automation, you can copy-paste the following snippet as a workflow into your repository:

```yaml
name: Project automation

on:
  issues:
    types:
      - opened

jobs:
  add_issue_to_project:
    name: Add new issues to project board
    runs-on: ubuntu-latest
    steps:
      - run: |
          gh api graphql -f query='mutation($project:ID!, $issue:ID!) { addProjectNextItem(input: {projectId: $project, contentId: $issue}) { projectNextItem { id }  } }' -f project=${{ secrets.PROJECT_BOARD_ID }} -f issue=${{ github.event.issue.node_id }}
        env:
          GITHUB_TOKEN: ${{ secrets.PROJECT_AUTOMATION_TOKEN }}
```

The requirements are:

1. Have a PAT as a secret named `PROJECT_AUTOMATION_TOKEN` that has **project write access**
2. Have a secret named `PROJECT_BOARD_ID` that has the *node_id* of the project you want to add the issue to

You will have to add this workflow to every repository where you want new issues to automatically be added to the project board.

**Reply by kobakou** [reply]
@thomaseizinger 

Thank you for giving your good solution! I've tried this. But I can't complete actions in success. This error is wired for me. When I tried this API call(mutation addProjectNextItem using issue.node_id) by hand, it result in success and added the issue into the org project.

Do you have any advice? 

This is a raw log of the action.
```
2022-03-05T10:29:19.1435775Z ##[group]Run gh api graphql -f query='mutation($project:ID!, $issue:ID!) { addProjectNextItem(input: {projectId: $project, contentId: $issue}) { projectNextItem { id }  } }' -f project=*** -f issue=I_kwDOBcRqB85FKOxO
2022-03-05T10:29:19.1437177Z [36;1mgh api graphql -f query='mutation($project:ID!, $issue:ID!) { addProjectNextItem(input: {projectId: $project, contentId: $issue}) { projectNextItem { id }  } }' -f project=*** -f issue=I_kwDOBcRqB85FKOxO[0m
2022-03-05T10:29:19.2179352Z shell: /usr/bin/bash -e {0}
2022-03-05T10:29:19.2179773Z env:
2022-03-05T10:29:19.2180552Z   GITHUB_TOKEN: ***
2022-03-05T10:29:19.2181182Z ##[endgroup]
2022-03-05T10:29:20.6546728Z gh: Could not resolve to a node with the global id of 'I_kwDOBcRqB85FKOxO'.
2022-03-05T10:29:20.6547988Z {"data":{"addProjectNextItem":null},"errors":[{"type":"NOT_FOUND","path":["addProjectNextItem"],"locations":[{"line":1,"column":38}],"message":"Could not resolve to a node with the global id of 'I_kwDOBcRqB85FKOxO'."}]}
2022-03-05T10:29:20.6576555Z ##[error]Process completed with exit code 1.
```

**Reply by kobakou** [reply]
@thomaseizinger 

thank you. I've resolved this issue! I assumed this is a token permission issue.

👎  error case: token created by web applications
👍  solved case: token created by gh/cli `gh auth login --scopes "write:org"`

thanks!!

**Reply by antgamdia** [reply]
Thanks for sharing the workflow, it works pretty well. In case it's useful for someone, let me share how I got the values for each env var:

#### `PROJECT_AUTOMATION_TOKEN` 

This is a GitHub personal access token (PAT), ideally should be a bot account. It is proven to work with `write:org,repo,workflow` scopes, but perhaps we can decrease them.
> Note that the tokens created with the gh CLI already request the `repo` and `workflow` scopes by default. That's why some users were getting `Could not resolve to a node with the global id of XXXX` if they just created a token with `write:org` scope from the UI.

In short, for getting a token with the proper scopes, just go to: https://github.com/settings/tokens/new?scopes=write:org,repo,workflow

#### `PROJECT_BOARD_ID` 

The easiest way is to install the GitHub CLI, login and then execute:

```
gh api graphql -f query='query{ organization(login: \"YOUR_ORGANIZATION_NAME\"){ projectNext(number: YOUR_PROJECT_NUMER) { id } } }'
```

> If you don't need to escape the slashes, just replace `\"` with  `"`.

For instance, if your project URL is `https://github.com/orgs/myreallycoolorganization/projects/3/views/1` you would need to run `gh api graphql -f query='query{ organization(login: \"myreallycoolorganization\"){ projectNext(number: 3) { id } } }'`

**Reply by seiyria** [reply]
How many project minutes does this burn per run? I'm a bit frustrated that we'd basically have to pay for a feature that should be built in, but hasn't had much traction in almost 2 years, but it's basically required for using Projects.

**Reply by antgamdia** [reply]
According to my latest runs, it takes 12s-18s per issue, but yep, I agree, it should be a built-in functionality for sure :S

**Reply by Stf-F** [reply]
@antgamdia , @seiyria , unless I am missing the point, adding a new issue to a repo now adds it to the Project (beta). You can also chose which column to map it to by default (New, Backlog etc.).
  
You will need to make sure to assign the issue to the project though for it to work, otherwise 🚫 .

![image](https://user-images.githubusercontent.com/52630104/179753444-4d255368-38e5-4b59-91bf-bb57699524d6.png)

**Reply by seiyria** [reply]
@Stf-F no, based on what I'm seeing that actually only fires when the issue is added TO the project, it doesn't cover adding the issue to the project. so this still stands.

**Reply by Stf-F** [reply]
@seiyria , not sure I got your point there? If I create an issue in my repo and assign it to the Project (screenie) while creating it, or with the batch feature available on the list view, the new issue ends up being visible to the Project in the chosen column . 

Is this not what you were looking for?

![Pasted_Image_19_07_2022__15_01](https://user-images.githubusercontent.com/52630104/179769406-56613c0e-a4f4-4957-8278-d28dfda30b2c.png)

**Reply by seiyria** [reply]
@Stf-F no, I want to automatically add issues to the project when they're created - that's what this entire discussion is about.

**Reply by petewalker** [reply]
@Stf-F - As per the roadmap issue - https://github.com/github/roadmap/issues/286

>Currently items need to be added manually, which works well for some use-cases but can be repetitive for projects that want to see _all_ the issues from a given repository.

**Comment by jluethi** [comment]
@thomaseizinger Your automation was working nicely for me for a while, but now it broke with the following error:
```
gh: The `ProjectNext` API is deprecated in favour of the more capable `ProjectV2` API. Follow the ProjectV2 guide at https://github.blog/changelog/2022-06-23-the-new-github-issues-june-23rd-update/, to find a suitable replacement.
```

Even though the [website here](https://docs.github.com/en/graphql/reference/mutations#addprojectnextitem) states it would only be deprecated next year, it already seems to be the case now.
Does anyone with a better understanding of the github `ProjectV2` API than me have a way to use the new API to achieve the same thing, i.e. adding new issues to the project automatically?

**Reply by lfrancke** [reply]
I believe a query along these lines would be the replacement:

```graphql
mutation($project:ID!, $issue:ID!) {
  addProjectV2ItemById(input: {projectId: $project, contentId: $issue}) {
    item {
      id
    }
  }
}
```

You can try over here: https://docs.github.com/en/graphql/overview/explorer

**Reply by jluethi** [reply]
Thanks @lfrancke !

I don't have experience with the github API, so some of it is a bit confusing to me. I replaced my old mutation query with this new one, but now get the following error:
```
gh: Could not resolve to ProjectV2 node with the global id of '***'.
{"data":{"addProjectV2ItemById":null},"errors":[{"type":"NOT_FOUND","path":["addProjectV2ItemById"],"locations":[{"line":2,"column":3}],"message":"Could not resolve to ProjectV2 node with the global id of '***'."}]}
```

with the query:
```
gh api graphql -f query='mutation($project:ID!, $issue:ID!) {
  addProjectV2ItemById(input: {projectId: $project, contentId: $issue}) {item {
      id
    }
  }
}' -f project=${{ secrets.PROJECT_BOARD_ID }} -f issue=${{ github.event.issue.node_id }}
```

I first thought the project IDs may be different for ProjectV2, but `gh api graphql -f query='query{ organization(login: \"YOUR_ORGANIZATION_NAME\"){ projectV2(number: YOUR_PROJECT_NUMER) { id } } }'` returns the same project ID as before. Has something about using the ID changed?


----

I've also tried it with the helpful explorer you linked to. But there, I get a `"message": "Your token has not been granted the required scopes to execute this query. The 'addProjectV2ItemById' field requires one of the following scopes: ['project'], but your token has only been granted the: ['read:discussion', 'read:enterprise', 'read:gpg_key', 'read:org', 'read:packages', 'read:project', 'read:public_key', 'read:repo_hook', 'repo', 'user'] scopes. Please modify your token's scopes at: https://github.com/settings/tokens."`. I'm signed in with my github account and have activated the oath access in my org. I'm not sure where to provide the correct access token for the explorer though

**Reply by jluethi** [reply]
I now switched to using https://github.com/marketplace/actions/add-to-github-projects and it's working nicely

**Comment by Mdshoyonmia** [comment]
3143 Al Basatin
6886
Ar Rimal
Riyadh 13438
SA
